app/models/prompt.rb

```diff
@@ -163,22 +163,29 @@ def allowed_tags
   validate :restricted_tags
   def restricted_tags
     restriction = prompt_restriction
-    if restriction
-      TagSet::TAG_TYPES_RESTRICTED_TO_FANDOM.each do |tag_type|
-        if restriction.send("#{tag_type}_restrict_to_fandom")
-          # tag_type is one of a set set so we know it is safe for constantize
-          allowed_tags = tag_type.classify.constantize.with_parents(tag_set.fandom_taglist).canonical
-          disallowed_taglist = tag_set ? eval("tag_set.#{tag_type}_taglist") - allowed_tags : []
-          # check for tag set associations
-          disallowed_taglist.reject! {|tag| TagSetAssociation.where(tag_id: tag.id, parent_tag_id: tag_set.fandom_taglist).exists?}
-          unless disallowed_taglist.empty?
-            errors.add(:base, ts("^These %{tag_label} tags in your %{prompt_type} are not in the selected fandom(s), %{fandom}: %{taglist} (Your moderator may be able to fix this.)",
-                              prompt_type: self.class.name.downcase,
-                              tag_label: tag_type_label_name(tag_type).downcase, fandom: tag_set.fandom_taglist.collect(&:name).join(ArchiveConfig.DELIMITER_FOR_OUTPUT),
-                              taglist: disallowed_taglist.collect(&:name).join(ArchiveConfig.DELIMITER_FOR_OUTPUT)))
-          end
+    return unless restriction
+
+    TagSet::TAG_TYPES_RESTRICTED_TO_FANDOM.each do |tag_type|
+      next unless restriction.send("#{tag_type}_restrict_to_fandom")
+
+      # tag_type is one of a set set so we know it is safe for constantize
+      allowed_tags = tag_type.classify.constantize.with_parents(tag_set.fandom_taglist).canonical
+      disallowed_taglist = tag_set ? tag_set.send("#{tag_type}_taglist") - allowed_tags : []
+
+      # check for tag set associations
+      tag_set_associations = tag_set.owned_tag_set&.tag_set_associations
+      if tag_set_associations
+        disallowed_taglist.reject! do |tag|
+          tag_set_associations.exists?(tag_id: tag.id, parent_tag_id: tag_set.fandom_taglist)
         end
```

Comment:
What if:

```suggestion
      if tag_set_associations
        disallowed_taglist = disallowed_taglist - tag_set_associations.where(tag_id: disallowed_taglist, parent_tag_id: tag_set.fandom_taglist).map(&:tag)
      end
```

The `reject!` is basically a loop, so it does one tag_set_associations exist query for every tag. My suggestion instead does one query to tag_set_associations and then for every tag to be removed it loads the tag. I think that's better because:
1) Loading the tag should be cheap because it uses the primary key on tags
2) There are naturally less tags to be removed than there are tags in disallowed_taglist. So the N for the N+1 query is smaller

Or maybe with the other comment above, this doesn't work at all. Or maybe there's a way around loading the tags. I'd just like to reduce the amount of N+1 queries going on here with all these loops.

Reply:
I think I've found a way to make it work!